scrapy_poet/utils.py

```diff
@@ -118,47 +120,57 @@ def is_response_going_to_be_used(request, spider):
     if is_callback_using_response(callback):
         return True
 
-    for provider in discover_callback_providers(callback):
+    providers = spider.settings["SCRAPY_POET_PROVIDERS"]
+    for provider in discover_callback_providers(callback, providers):
         if is_provider_using_response(provider):
             return True
 
     return False
 
 
-def build_plan(callback, instances) -> Tuple[andi.Plan, Dict[Type, PageObjectInputProvider]]:
+def build_plan(callback: Callable, providers: List[Type[PageObjectInputProvider]]) -> andi.Plan:
     """Build a plan for the injection in the callback."""
-    provider_instances = build_providers(instances)
-    plan = andi.plan(
+    return andi.plan(
         callback,
         is_injectable=is_injectable,
-        externally_provided=provider_instances.keys()
+        externally_provided=get_provided_classes_from_providers(providers),
     )
-    return plan, provider_instances
 
 
-def build_providers(instances) -> Dict[Type, PageObjectInputProvider]:
-    result = {}
-    for cls, provider in providers.items():
-        kwargs = andi.plan(
-            provider,
-            is_injectable=is_injectable,
-            externally_provided=instances.keys(),
-            full_final_kwargs=True,
-        ).final_kwargs(instances)
-        result[cls] = provider(**kwargs)  # type: ignore
-
-    return result
+def build_provider(provider: Type[PageObjectInputProvider],
+                   external_dependencies: Dict[Callable, Any]) -> PageObjectInputProvider:
+    kwargs = andi.plan(
+        provider,
+        is_injectable=is_injectable,
+        externally_provided=external_dependencies.keys(),
+        full_final_kwargs=True,
+    ).final_kwargs(external_dependencies)
+    return provider(**kwargs)  # type: ignore
 
 
 @inlineCallbacks
-def build_instances(plan: andi.Plan, providers):
-    """Build the instances dict from a plan."""
+def build_instances(plan: andi.Plan, providers: List[Type[PageObjectInputProvider]],
+                    external_dependencies: Dict[Callable, Any]):
+    """Build the instances dict from a plan including external dependencies."""
     instances = {}
-    for cls, kwargs_spec in plan:
-        if cls in providers:
-            instances[cls] = yield maybeDeferred_coro(providers[cls])
-        else:
+
+    # Build dependencies handled by registered providers
+    dependencies_set = set(cls for cls, kwargs_spec in plan.dependencies)
+    for provider in providers:
+        provided_classes = dependencies_set & provider.provided_classes
+        provided_classes -= instances.keys()  # ignore already provided types
+        if not provided_classes:
+            continue
+
+        provider_instance = build_provider(provider, external_dependencies)
+        results = yield maybeDeferred_coro(provider_instance, provided_classes)
+        instances.update(results)
```

Comment:
Wonder if we should check that `results.keys() - provided_classes == {}` and raise an exception otherwise. This will help if somebody writes a wrong provider returning the wrong object in the dict keys.

Reply:
I would go further and raise an exception whenever the returned class set is different than the set the provider was asked. What do you think?

Reply:
Yes, to me it makes sense. Will help the developer to spot problems. I don't see the case where returning something different in the keys would make sense.

Reply:
> I would go further and raise an exception whenever the returned class set is different than the set the provider was asked. What do you think?

If we do that we will loose a powerful behavior in my opinion which is:

- imagine a SplashScreenshotProvider
- this provider provides SplashScreenshot and ResponseData
- it has priority over the ResponseDataProvider

Now let's think in two possible scenarios:

- callback needs SplashScreenshot and ResponseData
- callback only needs ResponseData

If callback needs both, okay, we can proceed and build the two objects as usual.

If callback only needs ResponseData, our SplashScreenshotProvider could then return a blank set of instances. This will force our injection middleware to ask for the next provider that's able to provide this information, that in our case is the ResponseDataProvider.

What do you think about it?

Reply:
Yeah, this makes sense. Good example :+1: . But this is not incompatible with the idea I had in mind: the error should only be raised if any of the keys in the returned `dict` is not in `Provider.provided_clases`. An empty `dict` will always pass this check.

Reply:
Cool. Raising an exception whenever a provider returns classes that it was not supposed to provide.